src/mlpack/core/optimizers/trust_region_newton/trust_region_newton_impl.hpp

```diff
@@ -0,0 +1,214 @@
+/**
+ * @file trust_region_newton_impl.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of the Trust Region Newton Method.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_TRUST_REGION_NEWTON_TRUST_REGION_NEWTON_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_TRUST_REGION_NEWTON_TRUST_REGION_NEWTON_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "trust_region_newton.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename FunctionType>
+TrustRegionNewton<FunctionType>::TrustRegionNewton(
+    FunctionType& function,
+    const double minGradientNorm,
+    const size_t maxIterations,
+    const size_t maxConjugateIterations,
+    const double eta0,
+    const double eta1,
+    const double eta2,
+    const double sigma1,
+    const double sigma2,
+    const double sigma3) :
+    function(function),
+    minGradientNorm(minGradientNorm),
+    maxIterations(maxIterations),
+    maxConjugateIterations(maxConjugateIterations),
+    eta0(eta0),
+    eta1(eta1),
+    eta2(eta2),
+    sigma1(sigma1),
+    sigma2(sigma2),
+    sigma3(sigma3)
+{ /* Nothing to do. */ }
+
+template<typename FunctionType>
+void TrustRegionNewton<FunctionType>::ConjugateGradient(
+    FunctionType& function,
+    const double delta,
+    const arma::mat& iterate)
+{
+  // The initial gradient values.
+  r = -gradient;
+  arma::mat d = r;
+  arma::mat hessian;
+
+  s.zeros();
+  double rTr = arma::dot(r, r);
+
+  // Whether to optimize until convergence.
+  const bool optimizeUntilConvergence = (maxConjugateIterations == 0);
+
+  const double minGradientNormConjugate = 0.5 * arma::norm(gradient, 2);
+  for (size_t itNum = 0; optimizeUntilConvergence ||
+      (itNum != maxConjugateIterations); ++itNum)
+  {
+    // If ||r^i|| <= e_k||detla f(w^k)|| then output sk =s^(i) and stop.
+    if (arma::norm(r, 2) <= minGradientNormConjugate)
+      break;
+
+    if (itNum > 0)
+    {
+      const double rTrUpdate = arma::dot(r, r);
+      d *= (rTrUpdate / rTr);
+      d += r;
```

Comment:
Might be better to have these as one line, but I'm not sure how much of a difference it would make for Armadillo---I guess that all on one line means that this can be done in only one loop over `d` instead of 2.